keps/2724-topology-aware-schedling/README.md

```diff
@@ -0,0 +1,574 @@
+  choose the value which can accommodate the most pods.
+
+### Enforcing the assignment
+
+When the workload has the PodSet assignments which are non-homogenous (all Pods
+landing in the same topology) and is about to start we modify the corresponding
+PodTemplates in the Job object to inject the `kueue.x-k8s.io/topology`
+scheduling gate. For homogenous PodSet assignments it is enough to inject the
+corresponding nodeSelector in the PodTemplates.
+
+Then, there is a new component, called `TopologyUngater`, which is a Workload
+reconciler which lists all pods for a given TAS PodSet, and ensures that the
+pods in the expected number are un-gated to a given value.
+
+Along with the scheduling gate, to each pod template the
+`kueue.x-k8s.io/workload` and `kueue.x-k8s.io/podset` labels / annotations are
+added to facilitate quick lookup (List request) for all pods corresponding to
+the workload (and more specifically PodSetAssignment) by `TopologyAssigner`.
+
+The TopologyUngater watches for pod events which trigger the reconciliation
+loop. The reconciliations are batched by 1s periods to make sure multiple
+created pods at the similar time don't trigger too many reconciliations.
+
+In order to prevent ungating more pods as expected for a PodSet we consider to
+use the expectations mechanism. The expectations are set for when we are about
+to ungate a Pod. The expectation is fulfield if the Pod is observed as ungated
+or the ungating request fails. We hold ungating if there are pending ungatings
+within the PodSet.
+
+### Support for ReplicatedJobs in JobSet
+
+Currently, in the PodSet we just keep the number of Pods created for a given
+PodTemplate. However, in case of JobSet the pods are naturally split between
+instances of replicated Jobs.
+
+Thus, without any extra work for JobSet, if JobSet has more than one
+ReplicatedJob its Pods could thus be assigned by TAS to different topologies,
+making the assignment suboptimal.
+
+In order to improve accuracy of the support for ReplicatedJobs we do the
+following API adjustments:
+
+```golang
+type PodSet struct {
+  ...
+
+  // ReplicatedJobCount indicates the number of replicated Jobs being span by
+  // the PodSet. Each replicated Job is assumed to have the same number of Pods
+  // with the same template.
+  // Default: 1
+  ReplicatedJobCount *int32
+
+  // ReplicatedJobKeyLabel specifies the name of the label which indicates
+  // the specific Job instance among ReplicatedJobs.
+  ReplicatedJobKeyLabel *string
+}
+```
+
+In case of JobSet `ReplicatedJobKeyLabel` could be either
+`jobset.sigs.k8s.io/job-index` or `jobset.sigs.k8s.io/replicatedjob-name` as
+both will uniquely identify a specific Job among the set of ReplicatedJob.
+
+```golang
+type PodSetAssignment struct {
+
+  // ReplicatedJobKey indicates the value of the Pod label which is used to
+  // specify a specific Job among a set of ReplicatedJobs.
+  ReplicatedJobKey *string
```

Comment:
```suggestion
  ReplicatedJobLabelValue *string
```
?

Reply:
But how does Kueue know what the possible values are? or does it assume integers?

Reply:
Good question, with no trivial answer. Let me list options.

My original intention was to actually use  `ReplicatedJobKey` rather than `ReplicatedJobLabelValue`. In this approach I don't need to know the values, I only care that Pods which have the same value of the label belong to the same instance of Job. The TopologyUngater would need to keep mapping in cache from the value to the PodSetAssignment index (indicating the topology domain).

When TopologyUngater sees a Pod it would list all pods which belong to the same Workload, have the same PodSet name, and have the same value of `ReplicatedJobKey`. If the value is not yet in cache it will add it. Then, it will assign the Pods to the cached topology domain and ungate.  We can prevent race conditions with the expectations mechanism. In case of Kueue restart it will need to build the cache before ungating. I know this sounds complex, but should work, IIUC.

An alternative approach I see is to assign the values up-front when creating PodSetAssignments. This can be done since JobSet currently creates the Jobs while it is still suspended. However, I would prefer not to rely on this. First, it can be considered problematic: https://github.com/kubernetes-sigs/jobset/issues/535, second it would mean we need to wait until all Jobs are created.

Reply:
On second thought I think it makes sense to also include `ReplicatedJobLabelValue` in the status. It will cost another Workload update, but I think will greatly improve debuggability.

Reply:
Marking as resolved since we decided not to support the ReplicatedJob level for Alpha. Moved the started effort to alternatives.